### ISSUE TITLE
Retrieving min-cut from max-flow residual

### ISSUE THREAD

**Issue by manuel-castro** [scipy/scipy#14405]
Hello,

Is there a standard scipy way of retrieving the min-cut from the residual matrix returned by `scipy.sparse.csgraph.maximum_flow`?
Looking at the sparse graph routines, there doesn't seem to be a straightforward way of doing so, but I feel like I could be missing something.

Thank you in advance!!
Manuel

**Comment by czgdp1807** [comment]
For obtaining min-cut from maximum flow, I think running a DFS/BFS to find all the reachable nodes from source (including itself) will form one set of min-cut. The other is then self evident. Now, I think that we can introduce a function, `minimum_cut` which would first run `maximum_flow` and then DFS/BFS to obtain the sets of the cut and then return these as `MinimumCutResult`. It will help the user in getting the task done in one API call. Let me know if I understood your requirement correctly.
cc @fuglede @jjerphan @rgommers

**Comment by manuel-castro** [comment]
@czgdp1807 Yes exactly. I'm particularly interested in the edges removed from the graph, if those are returned in the `MinimumCutResult` that would be perfect

**Comment by fuglede** [comment]
For reference: The need for having a min cut also came up in #11498. It's a common enough operation that it would make sense to include explicitly.

Back then, I pondered if we'd want it in `MaximumFlowResult`, but could also make sense to keep it separately. Or taking the duality of the two problems seriously and providing a platform that elegantly handles both problems at the same footing.

It's worth spending some time to think through different use cases: Extending `MaximumFlowResult` puts constraints on what results possible future solvers should provide, which we may not want to do, which is an argument for not including it there. On the other hand, `MaximumFlowResult` exists only to allow adding new attributes while retaining backwards compatibility; if we knew we would never extend it, having a type instead of a tuple is a bit overkill; same would go for `MinimumCutResult`.

Finally, some methods might have an easier time providing one result over the other: If calculating a minimum cut fits naturally in the maximum flow solution, there's no reason to postpone the evaluation until afterwards. But then again, the translation is O(V), so cheap. Similarly, if one just wants a cut, maybe one would rather use a solver specialized for that, and avoid flows altogether.

In any case, you're right, @czgdp1807, simplest would be to just search from the source, following only non-saturated edges; this is what [networkx does as well](https://networkx.org/documentation/stable/_modules/networkx/algorithms/flow/maxflow.html#minimum_cut).

And just for reference, if someone comes across this thread at a point of time where nothing is available in SciPy yet, a (Cython friendly) solution would be the below. The only slight nuisance is that `MaximumFlowResult.residual` always includes edges in both directions, which we don't require the input to do, so we cheat a bit and use the non-public `csgraph._flow._add_reverse_edges` to make the input and residual matrices match up:

```python
from scipy.sparse.csgraph import maximum_flow
from scipy.sparse.csgraph._flow import _add_reverse_edges
from scipy.sparse import csr_matrix
import numpy as np

# Set up and example input; take the one from CLRS 2nd ed, p. 659
graph = csr_matrix([[0, 16, 13, 0, 0, 0],
                    [0, 0, 10, 12, 0, 0],
                    [0, 4, 0, 0, 14, 0],
                    [0, 0, 9, 0, 0, 20],
                    [0, 0, 0, 7, 0, 4],
                    [0, 0, 0, 0, 0, 0]]).astype(np.int32)
source = 0
sink = 5

# Calculate flow and add reverse edges
res = maximum_flow(graph, source, sink)
residual = res.residual
graph = _add_reverse_edges(graph)

# Arrays keeping track of results: cut contains the edges of
# a minimum cut; the value at an edge is true iff the edge
# is part of the cut. Similarly, in components, a vertex has a
# value of True iff it belongs to the component containing
# the source.
num_vertices = graph.shape[0]
num_edges = graph.data.shape[0]
cut = np.zeros(num_edges, dtype=bool)
components = np.zeros(num_vertices, dtype=bool)

# BFS from source, follow only unsaturated edges
q = np.empty(num_vertices, dtype=np.int32)
seen = np.zeros(num_vertices, dtype=bool)
q[0] = source
seen[0] = True
start = 0
end = 1
while start != end:
    vert = q[start]
    components[vert] = True
    start += 1
    for edge in range(graph.indptr[vert], graph.indptr[vert + 1]):
        if graph.data[edge] > 0:
            if graph.data[edge] == residual.data[edge]:
                cut[edge] = True
            else:
                head = graph.indices[edge]
                if seen[head]:
                    continue
                else:
                    seen[head] = True
                    q[end] = head
                    end += 1

# At this point, cut contains three edges; 1-3, 4-3, 4-5:
print(np.where(cut))  # (array([ 4, 14, 15]),)

# And components give the split in {0, 1, 2, 4}, {3, 5}:
print(components)  # [ True  True  True False  True False]

**Comment by manuel-castro** [comment]
Thank you @fuglede, that is very helpful! Think I'll go with your second approach until scipy implements a specialized min_cut function
